internal/pkg/agent/cmd/container.go

```diff
@@ -989,3 +994,152 @@ func isContainer(detail component.PlatformDetail) component.PlatformDetail {
+func shouldFleetEnroll(setupCfg setupConfig) (bool, error) {
+	if !setupCfg.Fleet.Enroll {
+		// Enrollment is explicitly disabled in the setup configuration.
+		return false, nil
+	}
+
+	if setupCfg.Fleet.Force {
+		// Enrollment is explicitly enforced by the setup configuration.
+		return true, nil
+	}
+
+	agentCfgFilePath := paths.AgentConfigFile()
+	_, err := statAgentConfigFile(agentCfgFilePath)
+	if os.IsNotExist(err) {
+		// The agent configuration file does not exist, so enrollment is required.
+		return true, nil
+	}
+
+	ctx := context.Background()
+	store, err := newEncryptedDiskStore(ctx, agentCfgFilePath)
+	if err != nil {
+		return false, fmt.Errorf("failed to instantiate encrypted disk store: %w", err)
+	}
+
+	reader, err := store.Load()
+	if err != nil {
+		return false, fmt.Errorf("failed to load from disk store: %w", err)
+	}
+
+	cfg, err := config.NewConfigFrom(reader)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedConfig, err := configuration.NewFromConfig(cfg)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedFleetHosts := storedConfig.Fleet.Client.GetHosts()
+	if len(storedFleetHosts) == 0 || !slices.Contains(storedFleetHosts, setupCfg.Fleet.URL) {
+		// The Fleet URL in the setup does not exist in the stored configuration, so enrollment is required.
+		return true, nil
+	}
+
+	// Evaluate the stored enrollment token hash against the setup enrollment token if both are present.
+	// Note that when "upgrading" from an older agent version the enrollment token hash will not exist
+	// in the stored configuration.
+	if len(storedConfig.Fleet.EnrollmentTokenHash) > 0 && len(setupCfg.Fleet.EnrollmentToken) > 0 {
+		enrollmentHashBytes, err := base64.StdEncoding.DecodeString(storedConfig.Fleet.EnrollmentTokenHash)
+		if err != nil {
+			return false, fmt.Errorf("failed to decode hash: %w", err)
+		}
+
+		err = crypto.ComparePBKDF2HashAndPassword(enrollmentHashBytes, []byte(setupCfg.Fleet.EnrollmentToken))
+		switch {
+		case errors.Is(err, crypto.ErrMismatchedHashAndPassword):
+			// The stored enrollment token hash does not match the new token, so enrollment is required.
+			return true, nil
+		case err != nil:
+			return false, fmt.Errorf("failed to compare hash: %w", err)
+		}
+	}
+
+	// Validate the stored API token to check if the agent is still authorized with Fleet.
+	log, err := logger.New("fleet_client", false)
+	if err != nil {
+		return false, fmt.Errorf("failed to create logger: %w", err)
+	}
+	fc, err := newFleetClient(log, storedConfig.Fleet.AccessAPIKey, storedConfig.Fleet.Client)
+	if err != nil {
+		return false, fmt.Errorf("failed to create fleet client: %w", err)
+	}
+
+	// Perform an ACK request with **empty events** to verify the validity of the API token.
+	// If the agent has been manually un-enrolled through the Kibana UI, the ACK request will fail due to an invalid API token.
```

Comment:
File an issue in https://github.com/elastic/ingest-docs to document this and other new behavior.

This makes so that if you want to unenroll an containerized agent, you can't do it from the UI in the normal way. You either need to revoke the enrollment token and/or remove the Fleet URL and Token from the environment variables.

Someone force uenrolling an agent in the UI now has the behavior that the next container restart will cause the same agent to re-enroll itself with a different ID.

Reply:
roger that! just to be sure, the behaviour of this PR is what we are after, right? In other words, should I revert this one and change it?

Reply:
The behavior is different for containers now, but I think in a way that is OK. We just need to document it. The user left the container with a valid enrollment token and essentially told it to enroll at startup if not enrolled, we just do a better job of detecting that.